src/timezones.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <port.h>
+#include <catalog/pg_type.h>
+#include <executor/spi.h>
+#include <utils/builtins.h>
+
+#include "timezones.h"
+
+static SPIPlanPtr savedPlanNormalizeTimezone = NULL;
+
+/*
+ * For the given TZ name or abbreviation return the normalized TZ name.
+ * For invalid TZ name NULL is returned.
+ * Examples:
+ * - normalized_timezone_name("Europe/Moscow") = "Europe/Moscow"
+ * - normalized_timezone_name("MSK") = "Europe/Moscow"
+ * - normalized_timezone_name("Europe/Ololondon") = NULL
+ */
+const char *
+ts_validate_and_normalize_timezone_name(const char *tz_name)
+{
+	int res;
+	bool isnull;
+	char *buffer;
+	Datum qvalues[] = { CStringGetTextDatum(tz_name) };
+
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "could not connect to SPI");
+
+	if (savedPlanNormalizeTimezone == NULL)
+	{
+		Oid argtypes[] = { TEXTOID };
+		savedPlanNormalizeTimezone = SPI_prepare("SELECT name FROM pg_catalog.pg_timezone_names "
+												 "WHERE name = $1 OR abbrev = $1 "
+												 "ORDER BY name LIMIT 1",
+												 1,
+												 argtypes);
+
+		if (savedPlanNormalizeTimezone == NULL)
+			elog(ERROR, "Error preparing query");
```

Comment:
Non-conforming error. Also, I think this requires a bit more context as it would be hard to understand what this means.

```suggestion
			elog(ERROR, "error preparing query");
```